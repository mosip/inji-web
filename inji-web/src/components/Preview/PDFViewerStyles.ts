export const PDFViewerStyles = {
    'container': 'w-full max-h-screen relative flex flex-col items-center overflow-auto',
    'pageWrapper': 'mb-2 border border-gray-200'
};