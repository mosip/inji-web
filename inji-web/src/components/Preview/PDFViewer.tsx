import {SpinningLoader} from "../Common/SpinningLoader";
import {useEffect, useMemo, useRef, useState} from "react";

import {Document, Page, pdfjs} from "react-pdf";
import {pdfWorkerSource} from "../../utils/constants";
import {PDFViewerStyles} from "./PDFViewerStyles";

pdfjs.GlobalWorkerOptions.workerSrc = pdfWorkerSource(pdfjs.version);

export function PDFViewer(props: Readonly<{ previewContent: Blob }>) {
    const blobUrl = useMemo(() => URL.createObjectURL(props.previewContent), [props.previewContent]);

    const containerRef = useRef<HTMLDivElement>(null);
    const [containerWidth, setContainerWidth] = useState(0);
    const [numPages, setNumPages] = useState<number>(0);
    // Padding to account for any margins or borders in the container
    const containerPadding = 22;

    useEffect(() => {
        if (!containerRef.current) return;
        const observer = new ResizeObserver(entries => {
            for (let entry of entries) {
                setContainerWidth(entry.contentRect.width - containerPadding);
            }
        });
        observer.observe(containerRef.current);
        return () => {
            // cleanup the observer and URL
            observer.disconnect();
            URL.revokeObjectURL(blobUrl);
        }
    }, []);

    function onDocumentLoadSuccess({numPages}: { numPages: number }) {
        setNumPages(numPages);
    }

    return (
        <div ref={containerRef} className={PDFViewerStyles.container}>
            <Document
                file={blobUrl}
                onLoadSuccess={onDocumentLoadSuccess}
                renderMode="canvas"
                loading={<SpinningLoader/>}
            >
                {Array.from({length: numPages}, (_, i) => (
                    <div
                        key={`page-${i + 1}`}
                        className={PDFViewerStyles.pageWrapper}
                    >
                        <Page
                            key={i + 1}
                            pageNumber={i + 1}
                            width={containerWidth}
                            renderTextLayer={false}
                            renderAnnotationLayer={false}
                        />
                    </div>
                ))}
            </Document>
        </div>
    );
}