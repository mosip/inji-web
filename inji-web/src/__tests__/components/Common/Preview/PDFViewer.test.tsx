import React from 'react';
import {render, screen} from '@testing-library/react';
import '@react-pdf-viewer/core/lib/styles/index.css';
import {PDFViewer} from '../../../../components/Preview/PDFViewer';

jest.mock('@react-pdf-viewer/core', () => ({
    SpecialZoomLevel: {
        PageWidth: 'page-width'
    },
    Worker: ({children}: { children: React.ReactNode }) => <div data-testid="pdf-worker">{children}</div>,
    Viewer: ({fileUrl, renderLoader, defaultScale, enableSmoothScroll}: {
        fileUrl: string;
        renderLoader?: (props: unknown) => React.ReactNode;
        defaultScale: string;
        enableSmoothScroll?: boolean
    }) => (
        <div
            data-testid="pdf-viewer"
            data-file-url={fileUrl}
            data-default-scale={defaultScale}
            data-enable-smooth-scroll={enableSmoothScroll ? 'true' : 'false'}
        >
            {/* Render a mock loader to test the renderLoader prop */}
            {renderLoader && renderLoader({})}
        </div>
    )
}));

jest.mock('../../../../components/Common/SpinningLoader', () => ({
    SpinningLoader: () => <div data-testid="spinning-loader">Loading PDF...</div>
}));

describe('PDFViewer Component', () => {
    const mockPdfUrl = 'https://example.com/test.pdf';

    it('renders correctly and matches snapshot', () => {
        const {container} = render(<PDFViewer previewContent={mockPdfUrl}/>);
        expect(container).toMatchSnapshot();
    });

    it('renders the PDF Worker component', () => {
        render(<PDFViewer previewContent={mockPdfUrl}/>);
        expect(screen.getByTestId('pdf-worker')).toBeInTheDocument();
    });

    it('renders the PDF Viewer component with correct props', () => {
        render(<PDFViewer previewContent={mockPdfUrl}/>);

        const viewer = screen.getByTestId('pdf-viewer');
        expect(viewer).toBeInTheDocument();
        expect(viewer).toHaveAttribute('data-file-url', mockPdfUrl);
        expect(viewer).toHaveAttribute('data-default-scale', 'page-width');
        expect(viewer).toHaveAttribute('data-enable-smooth-scroll', 'true');
    });

    it('passes the SpinningLoader component as renderLoader prop', () => {
        render(<PDFViewer previewContent={mockPdfUrl}/>);

        // Check if the spinning loader is rendered
        expect(screen.getByTestId('spinning-loader')).toBeInTheDocument();
        expect(screen.getByText('Loading PDF...')).toBeInTheDocument();
    });

    it('passes the correct URL to the Viewer component', () => {
        const customUrl = 'https://example.com/custom.pdf';
        render(<PDFViewer previewContent={customUrl}/>);

        const viewer = screen.getByTestId('pdf-viewer');
        expect(viewer).toHaveAttribute('data-file-url', customUrl);
    });
});