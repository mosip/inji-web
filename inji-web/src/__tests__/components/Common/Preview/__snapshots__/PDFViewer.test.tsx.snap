// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDFViewer Component renders correctly and matches snapshot 1`] = `
<div>
  <div
    data-testid="pdf-worker"
  >
    <div
      data-default-scale="page-width"
      data-enable-smooth-scroll="true"
      data-file-url="https://example.com/test.pdf"
      data-testid="pdf-viewer"
    >
      <div
        data-testid="spinning-loader"
      >
        Loading PDF...
      </div>
    </div>
  </div>
</div>
`;
